server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestCreateIndexAction.java

```diff
@@ -64,18 +64,36 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
         }
 
         CreateIndexRequest createIndexRequest = new CreateIndexRequest(request.param("index"));
+
         if (request.hasContent()) {
-            Map<String, Object> sourceAsMap = XContentHelper.convertToMap(request.content(), false, request.getXContentType()).v2();
-            if (includeTypeName == false && sourceAsMap.containsKey("mappings")) {
-                Map<String, Object> newSourceAsMap = new HashMap<>(sourceAsMap);
-                newSourceAsMap.put("mappings", Collections.singletonMap(MapperService.SINGLE_MAPPING_NAME, sourceAsMap.get("mappings")));
-                sourceAsMap = newSourceAsMap;
-            }
+            Map<String, Object> sourceAsMap = XContentHelper.convertToMap(request.requiredContent(), false,
+                request.getXContentType()).v2();
+            sourceAsMap = prepareMappings(sourceAsMap, includeTypeName);
             createIndexRequest.source(sourceAsMap, LoggingDeprecationHandler.INSTANCE);
         }
+
         createIndexRequest.timeout(request.paramAsTime("timeout", createIndexRequest.timeout()));
         createIndexRequest.masterNodeTimeout(request.paramAsTime("master_timeout", createIndexRequest.masterNodeTimeout()));
         createIndexRequest.waitForActiveShards(ActiveShardCount.parseString(request.param("wait_for_active_shards")));
         return channel -> client.admin().indices().create(createIndexRequest, new RestToXContentListener<>(channel));
     }
+
+
+    static Map<String, Object> prepareMappings(Map<String, Object> source, boolean includeTypeName) {
+        if (includeTypeName || !source.containsKey("mappings")) {
+            return source;
+        }
+
+        Map<String, Object> newSource = new HashMap<>(source);
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> mappings = (Map<String, Object>) source.get("mappings");
```

Comment:
This cast is unsafe and would result in a confusing error message if a user created an index with something like below? Should we have an `instanceof Map` check?

```
PUT test
{
  "mappings": "foo"
}
```

Reply:
Just a note that I made this method more robust, but after testing I noticed the same confusing error will occur even without these changes because of another unchecked cast: https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java#L381

Reply:
Thanks for looking